nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_common/connections/udp.hpp

```diff
@@ -0,0 +1,383 @@
+
+    ip_mreq mreq;
+    mreq.imr_multiaddr.s_addr = inet_addr(group_ip.c_str());  // Multicast group address
+    mreq.imr_interface.s_addr = inet_addr(host_.ip.c_str());
+
+    int result = setsockopt(sock_fd_, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, sizeof(mreq));
+    if (result < 0) throw SocketError(errno);
+
+    multicast_ip_.emplace(group_ip);
+    return *this;
+  }
+
+  /**
+   * @brief Bind the socket to host IP and port given in `init()`. If `join_multicast_group()` was
+   * called before this function, the socket will be bound to `group_ip` instead. At least `init()`
+   * has to have been called before.
+   */
+  UdpSocket & bind()
+  {
+    if (state_ < State::INITIALIZED) throw UsageError("Socket has to be initialized first");
+    if (state_ >= State::BOUND) throw UsageError("Re-binding already bound socket");
+
+    sockaddr_in addr{};
+    addr.sin_family = AF_INET;
+    addr.sin_port = htons(host_.port);
+    addr.sin_addr.s_addr = multicast_ip_ ? inet_addr(multicast_ip_->c_str()) : INADDR_ANY;
+
+    int result = ::bind(sock_fd_, (struct sockaddr *)&addr, sizeof(addr));
+    if (result == -1) throw SocketError(errno);
+
+    state_ = State::BOUND;
+    return *this;
+  }
+
+  /**
+   * @brief Register a callback for processing received packets and start the receiver thread. The
+   * callback will be called for each received packet, and will be executed in the receive thread.
+   * Has to be called on a bound socket (`bind()` has to have been called before).
+   *
+   * @param callback The function to be executed for each received packet.
+   */
+  UdpSocket & subscribe(callback_t && callback)
+  {
+    if (state_ < State::BOUND) throw UsageError("Socket has to be bound first");
+    if (state_ > State::BOUND) throw UsageError("Cannot re-subscribe to socket");
+
+    callback_ = std::move(callback);
+    launch_receiver();
+    return *this;
+  }
+
+  /**
+   * @brief Gracefully stops the active receiver thread (if any) but keeps the socket alive. The
+   * same socket can later be subscribed again.
+   */
+  UdpSocket & unsubscribe()
+  {
+    if (state_ == State::ACTIVE) state_ = State::BOUND;
+    if (receive_thread_.joinable()) receive_thread_.join();
+    return *this;
+  }
+
+  ~UdpSocket()
+  {
+    unsubscribe();
+    close(sock_fd_);
+  }
+
+private:
+  void launch_receiver()
+  {
+    assert(state_ == State::BOUND);
+    assert(callback_);
+
+    state_ = State::ACTIVE;
+    receive_thread_ = std::thread([this]() {
+      std::vector<uint8_t> buffer;
+      while (state_ == State::ACTIVE) {
+        auto data_available = is_data_available();
+        if (!data_available.has_value()) throw SocketError(data_available.error());
+        if (!data_available.value()) continue;
+
+        buffer.resize(buffer_size_);
+        auto msg_header = make_msg_header(buffer);
+
+        ssize_t recv_result = recvmsg(sock_fd_, &msg_header.msg, MSG_TRUNC);
+        if (recv_result < 0) throw SocketError(errno);
```

Comment:
It might be nice to handle `recv_result == 0` for when the other side closed the connection.

Reply:
This would only apply to TCP, UDP is connectionless and so `0` indicates a 0-length datagram in a blocking socket. Added a clarifying comment in 044208b.